src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -478,38 +493,63 @@ private ScoreOrderedResultRetriever(CloseableIterator<PrimaryKeyWithSortKey> sco
             this.executionController = executionController;
             this.queryContext = queryContext;
 
-            this.keysSeen = new HashSet<>();
-            this.updatedKeys = new HashSet<>();
+            this.processedKeys = new HashSet<>(limit);
+            this.pendingRows = new ArrayDeque<>(limit);
+            this.limit = limit;
         }
 
         @Override
         public UnfilteredRowIterator computeNext()
         {
-            @SuppressWarnings("resource")
-            UnfilteredRowIterator iterator = nextRowIterator(this::nextSelectedKeyInRange);
+            if (pendingRows.isEmpty())
+                fillPendingRows();
+            returnedRowCount++;
             // Because we know ordered keys are fully qualified, we do not iterate partitions
-            return iterator != null ? iterator : endOfData();
+            return !pendingRows.isEmpty() ? pendingRows.poll() : endOfData();
         }
 
         /**
-         * Tries to obtain a row iterator for one of the supplied keys by repeatedly calling
-         * {@link ResultRetriever#apply} until it gives a non-null result.
-         * The keySupplier should return the next key with every call to get() and
-         * null when there are no more keys to try.
-         *
-         * @return an iterator or null if all keys were tried with no success
+         * Fills the pendingRows queue to generate a queue of row iterators for the supplied keys by repeatedly calling
+         * {@link #readAndValidatePartition} until it gives enough non-null results.
          */
-        private @Nullable UnfilteredRowIterator nextRowIterator(@Nonnull Supplier<PrimaryKeyWithSortKey> keySupplier)
+        private void fillPendingRows()
         {
-            UnfilteredRowIterator iterator = null;
-            while (iterator == null)
+            int count = Math.max(1, limit - returnedRowCount);
+            // Sort the primary keys by PrK order, just in case that helps with cache and disk efficiency
+            var keys = new TreeMap<PrimaryKey, List<PrimaryKeyWithSortKey>>();
+            // We want to get the first unique `count` keys to materialize
+            fillKeys(keys, count);
```

Comment:
I wonder if we know the list of keys in advance, then using a TreeMap is not the most efficient way to sort them. 
It creates additional heap allocations for each key. 
Why can't we just use `Collections.sort` or similar? Or maybe that doesn't matter much, because limits are usually low. But anyways, I like to not create unnecessary garbage, especially per-key / per-row.

// edit:
Oh, I can see we "refill" the keys whenever we get a key that didn't pass validation.
So maybe a `PriorityQueue` ? We have some nice optimized cheap to construct `PriorityQueue` already.